waiter/src/waiter/handler.clj

```diff
@@ -1174,14 +1174,52 @@
+  Handles Post /drain requests and sets the drain-until for the router based on the provided 'drain-timeout-ms' query
+  param which defaults to 120 seconds. This will cause the /status health check to fail and remove the router in general
+  load balancers. The request may have the 'crash' query parameter, which when true, will cause the router to exit after
+  the provided timeout. You can cancel the 'crash' request by sending a subsequent request setting it back to false before
+  the 'drain-timeout-ms' is reached."
+  [clock drain-atom admin-user?-fn crash-fn drain-mode?-fn {:keys [request-method] :as request}]
+  (try
+    (case request-method
+      :get (utils/clj->json-response {:result @drain-atom
+                                      :drain-mode? (drain-mode?-fn)})
+      :post (let [{:strs [drain-timeout-ms] :or {drain-timeout-ms "120000"} :as request-params} (-> request ru/query-params-request :query-params)
+                  auth-user (get request :authorization/user)
+                  drain-timeout-ms (utils/parse-int drain-timeout-ms)
+                  crash-process? (utils/request-flag request-params "crash")
+                  now (clock)
+                  drain-until (t/plus now (t/millis drain-timeout-ms))]
+              (when (not (admin-user?-fn auth-user))
+                (throw (ex-info "Must be an admin to use this endpoint."
+                                {:auth-user auth-user :log-level :info :status http-401-unauthorized})))
+              (log/warn "Putting router into drain mode! It should begin failing health checks.")
+              (swap! drain-atom assoc :drain-until drain-until :crash-process? crash-process?)
+              (when crash-process?
+                (log/warn "Going to attempt to kill router process after timeout." {:drain-timeout-ms drain-timeout-ms})
+                (async/go
+                  (async/<! (async/timeout drain-timeout-ms))
+                  (if (:crash-process? @drain-atom)
+                    (crash-fn)
+                    (log/warn "Cancelled attempt to kill waiter process!"))))
+              (utils/clj->json-response {:result @drain-atom}))
+      (throw (ex-info "Unsupported request method" {:log-level :info :method request-method :status http-405-method-not-allowed})))
+    (catch Exception ex
+      (utils/exception->response ex request))))
+
 (defn status-handler
   "Responds with an 'ok' status.
-   Includes representation of request if requested using the include=request-info query param."
-  [{:keys [body trailers-fn] :as request}]
+   Includes representation of request if requested using the include=request-info query param.
+   If the router is in drain mode then return 503 server error."
+  [drain-mode?-fn {:keys [body trailers-fn] :as request}]
   (try
     (when (instance? InputStream body)
       (log/info "consuming request body before rendering response")
       (slurp body))
+    (when (drain-mode?-fn)
```

Comment:
We should report [503 Service Unavailable](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503)

Reply:
This will be a frequently used endpoint.

Reply:
You even said `503` in the comment... 😞